src/DistrictResources/DistrictResources.cpp

```diff
@@ -1,4 +1,277 @@
+#include <fstream>
+#include <sstream>
+#include <iostream>
+#include <valarray>
 #include "DistrictResources.h"
+#include "../Vehicle/FireEngine.h"
+#include "../Vehicle/FireLadder.h"
+#include "../Vehicle/Vehicle.h"
+#include "../Event/Event.h"
 
 namespace LockFreeDispatch {
+
+    void DistrictResources::populateDistrictVehicles(const std::vector<Vehicle>& vehicles)
+    {
+        for (auto vehicle : vehicles)
+        {
+            districtVehicles.push_back(vehicle);
+        }
+    }
+
+    void DistrictResources::populateDistrictFireStations(const std::vector<FireStation>& stations)
+    {
+        for (auto station : stations)
+        {
+            districtFireStations.push_back(station);
+        }
+    }
+
+    FireStation DistrictResources::findFireStation(uint32_t fireStationID) {
+        for (auto station : districtFireStations)
+        {
+            if (station.getFireStationID() == fireStationID)
+            {
+                return station;
+            }
+        }
+        return {};
+    }
+
+    void DistrictResources::processVehicleSampleData(const std::string& fileName)
+    {
+        std::vector<std::vector<std::string>> vehicles;
+        std::vector<std::string> row;
+        std::string line, word;
+
+        std::fstream file (fileName, std::ios::in);
+        while(getline(file, line))
+        {
+            row.clear();
+
+            std::stringstream str(line);
+
+            while(getline(str, word, ','))
+                row.push_back(word);
+            vehicles.push_back(row);
+        }
+
+        bool header = true;
+
+        for (std::vector<std::string> entry : vehicles)
+        {
+            if (header)
+            {
+                header = false;
+                continue;
+            }
+            std::string vehicleType = entry.at(1);
+            Vehicle newVehicle{};
+            if (vehicleType == "Engine")
+            {
+                newVehicle = *new FireEngine();
+            }
+            else if (vehicleType == "Ladder")
+            {
+                newVehicle = *new FireLadder();
+            }
+
+            newVehicle.setVehicleID(stoi(entry.at(0)));
+            newVehicle.setHomeFireStation(findFireStation(stoi(entry.at(2))));
+            newVehicle.setMaxNumCrew(stoi(entry.at(3)));
+            newVehicle.setCurNumCrew(stoi(entry.at(4)));
+            newVehicle.setMaxWaterVolumeLitres(stof(entry.at(5)));
+            newVehicle.setCurWaterVolumeLitres(stof(entry.at(6)));
+            newVehicle.setWorkFactor(stof(entry.at(7)));
+
+            std::cout << "Adding vehicle #" << newVehicle.getVehicleID() << " to global list" <<std::endl;
+            districtVehicles.push_back(newVehicle);
```

Comment:
**I see now**. You are using pointers for polymorphism. A switch to values+references is not appropriate. However, I don't think this code will behave the way you are expecting, as the `std::vector` will copy the objects with a static cast into the base class. (See _object slicing_: https://stackoverflow.com/a/274634/2769271)

To preserve the original (derived class) objects, you need to copy the addresses, not the objects, into the `std::vector` so that you can dynamically cast them back.  For this, you will almost certainly want to use "smart pointers" to auto-deallocate, i.e., change the type of `districtVehicles` to `std::vector< std::unique_ptr< Vehicle > >`, or you are more likely to have memory leaks.


```suggestion
            auto const vehicle_id = stoi( entry.at( 0 ) );
            auto const home_fire_station = findFireStation( stoi( entry.at( 2 ) ) );
            // etc.
            if (vehicleType == "Engine")
            {
                districtVehicles.push_back( std::make_unique( FireEngine( vehicle_id, home_fire_station/*, etc.*/) ) );
            }
            else if (vehicleType == "Ladder")
            {
                // as above
            }

            std::cout << "Added vehicle #" << districtVehicles.back().getVehicleID() << " to global list" <<std::endl;
```
Note, this will probably introduce breaking changes to anywhere you use the `districtVehicles` member variable, but I think that code probably didn't perform correctly.

Reply:
@sean-chester Why did you use a `unique_ptr` instead of a `shared_ptr`. I admit my knowledge of C++ smart pointers is limited so I may be missing some key points, but having spent some time today trying to work with these non-shared pointers, I think I may be better suited with the shared variety.

There are several instances where I attempt to make a copy of the `districtVehicles` vector, such as when determining which vehicles meet an event's requirements or when returning a vector of `Vehicles` ordered by proximity to an event. From my perspective, using `unique_ptrs` makes that task significantly more difficult, if not impossible. Is my thinking flawed in that regard? If so, please feel free to correct me as I agree that smart pointers are a tool I should use but am struggling to properly implement them.

Reply:
@ethan-mcnamara , my recommendation is to proceed with `shared_ptr` and mark conversion to `unique_ptr` as a todo so that this isn't blocking.

I used a `unique_ptr` because (to the extent that I understand the code) there is only one object that should "own" the `Vehicle` object, i.e., be responsible for deleting it. (See Section 3.1 of [this article on resource management in cpp](https://www.stroustrup.com/resource-model.pdf).)  A `shared_ptr` is useful when that responsibility is best described by the policy "the last person to leave has to take out the trash," i.e., no object is specifically responsible for deleting another object, but will be if no other co-owners are left. That pattern is, in my opinion, pretty rare, but the programming is simpler.

For completeness, raw pointers are still valuable when you just want an address, not to denote ownership. More often than not, references will do, but sometimes you might want a vector of addresses but that vector is not responsible for deleting the objects it points to.

The challenge that you are running into is that when you copy a `districtVehicles` object, you also copy all of the ownership semantics associated with that vector. That isn't what you intend to do, I think. Probably, you want to create a vector of raw pointers to vehicle objects rather than copying them.

However, this could be a rabbit hole best reserved for a less deadline-sensitive period. It is an example of a general programming pattern called [Resource Acquisition Is Initialisation (RAII)](https://stackoverflow.com/a/18054738/2769271).

Reply:
@sean-chester, after doing some further reading (thank you for the references) I think I may have discovered my mistake and I believe I now understand your original intention of the use of the `unique_ptr`. I believe the issues I was experiencing attempting to copy the `districtVehicles` were due to me creating a variable of type `std::vector<std::unique_ptr<Vehicle>>` and attempting to copy into it the contents of `districtVehicles`, which is the same data type. This was impossible, as I the `unique_ptr` type does not allow copies to be made. 

I had replaced nearly every instance of `std::vector<Vehicle>` with `std::vector<std::unique_ptr<Vehicle>>`, as I did not properly understand the purpose of the `unique_ptr` class and believed it a universal replacement for the raw pointer, which I now realize is not its purpose.

I see now the flaw in this approach, as the goal of the `unique_ptr` is to stop this ownership of a `Vehicle` by multiple objects. I believe a better approach would be to keep `districtVehicles` as a `std::vector<std::unique_ptr<Vehicle>>`, but have all other (non-globally accessible) objects who may wish to copy this `vector` be of type `std::vector<Vehicle*>`, and to use the `std::unique_ptr::get()` function to access a pointer to the `Vehicle` object. 

I believe this would maintain a single-object ownership while allowing other objects to create filtered copies of the global `vector`, which is the task with which I was struggling yesterday.

Reply:
@ethan-mcnamara , precisely. I'm surprised that you connected all that together so quickly. Your proposed solution is, in my opinion, the best.